File: app/src/main/java/ru/gidravpn/hydra/data/subscription/WireGuardParser.kt
package ru.gidravpn.hydra.data.subscription

import android.net.Uri
import android.util.Base64
import org.json.JSONObject
import ru.gidravpn.hydra.data.model.Protocol

/**
 * Разбор конфигов WireGuard / AmneziaWG.
 *
 * Поддерживает:
 *  - стандартный `.conf` (секции [Interface]/[Peer]) — в т.ч. с параметрами
 *    обфускации AmneziaWG (Jc, Jmin, Jmax, S1, S2, H1–H4 — версии 1.0/1.5;
 *    I1–I5 — версия 2.0);
 *  - ссылки `wireguard://<base64 конфига>` и `awg://<base64 конфига>`
 *    (де-факто стандарт экосистемы: конфиг кодируется целиком).
 *
 * Ключи и параметры складываются в [ServerProfile.extra] как JSON —
 * см. [WireGuardConfigBuilder], который собирает из них .conf / uapi.
 */
object WireGuardParser {

    /**
     * @return карта полей extra-JSON профиля либо null, если вход не похож на WG/AWG.
     */
    fun parse(text: String, isAmnezia: Boolean): Pair<Map<String, String>, JSONObject>? {
        val conf = maybeUnwrapLink(text) ?: return null
        return parseConf(conf, isAmnezia)
    }

    /** Разворачивает ссылку wireguard://…/awg://… в текст .conf. */
    private fun maybeUnwrapLink(text: String): String? {
        val t = text.trim()
        return when {
            t.startsWith("wireguard://", ignoreCase = true) -> decodeBody(t.substringAfter("://"))
            t.startsWith("wg://", ignoreCase = true) -> decodeBody(t.substringAfter("://"))
            t.startsWith("awg://", ignoreCase = true) -> decodeBody(t.substringAfter("://"))
            t.startsWith("amnezia://", ignoreCase = true) -> decodeBody(t.substringAfter("://"))
            "[Interface]" in t -> t
            else -> null
        }
    }

    private fun decodeBody(body: String): String? = runCatching {
        val cleaned = body.substringBefore("#").replace('-', '+').replace('_', '/')
        val pad = if (cleaned.length % 4 == 0) 0 else 4 - cleaned.length % 4
        String(Base64.decode(cleaned + "=".repeat(pad), Base64.DEFAULT))
    }.getOrElse { runCatching { Uri.decode(body) }.getOrNull() }

    /** Разбор .conf: первая секция [Interface] + первый [Peer]. */
    private fun parseConf(conf: String, isAmnezia: Boolean): Pair<Map<String, String>, JSONObject>? {
        if ("[Interface]" !in conf && "[Peer]" !in conf) return null

        val extra = JSONObject()
        val iface = mutableMapOf<String, String>()
        val peer = mutableMapOf<String, String>()
        var section = ""

        conf.lineSequence().forEach { raw ->
            val line = raw.trim()
            when {
                line.equals("[Interface]", true) -> section = "iface"
                line.equals("[Peer]", true) -> section = "peer"
                line.isEmpty() || line.startsWith("#") -> Unit
                section.isNotEmpty() && "=" in line -> {
                    val k = line.substringBefore("=").trim().lowercase()
                    val v = line.substringAfter("=").trim()
                    (if (section == "iface") iface else peer)[k] = v
                }
            }
        }

        val pubKey = peer["publickey"] ?: return null

        // Endpoint → address:port профиля; при отсутствии — хост-заглушка.
        val endpoint = peer["endpoint"].orEmpty()
        val host = endpoint.substringBeforeLast(":").removePrefix("[").removeSuffix("]")
        val port = endpoint.substringAfterLast(":").toIntOrNull() ?: 51820

        iface["privatekey"]?.let { extra.put("private_key", it) }
        iface["address"]?.let { extra.put("local_address", it) }
        iface["dns"]?.let { extra.put("dns", it) }
        iface["mtu"]?.let { extra.put("mtu", it) }
        peer["presharedkey"]?.let { extra.put("preshared_key", it) }
        peer["allowedips"]?.let { extra.put("allowed_ips", it) }
        peer["persistentkeepalive"]?.let { extra.put("keepalive", it) }

        // --- Обфускация AmneziaWG ---
        var hasAwg = false
        val awgParams = listOf(
            "jc", "jmin", "jmax", "s1", "s2", "h1", "h2", "h3", "h4",
            "i1", "i2", "i3", "i4", "i5"
        )
        awgParams.forEach { k ->
            iface[k]?.let { extra.put(k, it); hasAwg = true }
        }
        // Версия: 2.0 определяется по маркерам I1–I5, иначе 1.x
        val version = when {
            !hasAwg -> "plain"
            iface.keys.any { it in listOf("i1", "i2", "i3", "i4", "i5") } -> "2.0"
            else -> "1.0"
        }
        extra.put("awg_version", version)

        val meta = mapOf(
            "host" to host.ifEmpty { "" },
            "port" to port.toString(),
            "public_key" to pubKey
        )
        if (!isAmnezia && hasAwg) {
            // конфиг AmneziaWG, но пользователь импортировал как plain WG — не теряем параметры
        }
        return meta to extra
    }

    /** Сборка ServerProfile из разобранного конфига. */
    fun toProfile(text: String, fallbackName: String, isAmnezia: Boolean): ru.gidravpn.hydra.data.model.ServerProfile? {
        val (meta, extra) = parse(text, isAmnezia) ?: return null
        val proto = if (isAmnezia) Protocol.AMNEZIAWG else Protocol.WIREGUARD
        return ru.gidravpn.hydra.data.model.ServerProfile(
            name = fallbackName,
            protocolId = proto.id,
            address = meta["host"].orEmpty(),
            port = meta["port"]?.toIntOrNull() ?: 51820,
            uuidOrPassword = extra.optString("private_key"),   // приватный ключ интерфейса
            extra = extra.toString(),
        )
    }
}

File: ios/Packages/HydraKit/Sources/HydraKit/WireGuardParser.swift
import Foundation

/// Разбор конфигов WireGuard / AmneziaWG — порт `WireGuardParser` с Android: `.conf`
/// (секции [Interface]/[Peer]) и ссылки `wireguard://` / `awg://` с конфигом в base64.
/// Поля складываются в `extra` под теми же ключами, что на Android.
public enum WireGuardParser {

    /// Ключи обфускации AmneziaWG всех поколений 1.0 – 3.x (строчными, как в `AwgParams.ALL`).
    public static let awgKeys = [
        "jc", "jmin", "jmax", "s1", "s2", "s3", "s4", "h1", "h2", "h3", "h4", "i1", "i2", "i3", "i4", "i5",
        "headerprotectionkey", "contentpaddingaddition", "rekeyaftertime", "rekeytimeout", "rejectaftertime",
        "keepalivetimeout", "maxhandshakeattempts", "randomtrailers", "disablecookies",
    ]
    private static let v3Keys: Set<String> = [
        "headerprotectionkey", "contentpaddingaddition", "rekeyaftertime", "rekeytimeout", "rejectaftertime",
        "keepalivetimeout", "maxhandshakeattempts", "randomtrailers", "disablecookies",
    ]

    public static func toProfile(_ text: String, fallbackName: String, isAmnezia: Bool) -> ServerProfile? {
        guard let conf = unwrap(text), let parsed = parseConf(conf) else { return nil }
        let (host, port, publicKey, extra) = parsed
        var e = extra
        e["public_key"] = publicKey
        return ServerProfile(
            name: fallbackName,
            protocolId: (isAmnezia ? ServerProtocol.amneziaWG : .wireguard).rawValue,
            address: host, port: port,
            uuidOrPassword: (extra["private_key"] as? String) ?? "",
            extra: LinkParser.json(e))
    }

    static func unwrap(_ text: String) -> String? {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        for prefix in ["wireguard://", "wg://", "awg://", "amnezia://"] where t.lowercased().hasPrefix(prefix) {
            let body = String(t.dropFirst(prefix.count)).components(separatedBy: "#")[0]
            if let d = LinkParser.base64Decode(body), let s = String(data: d, encoding: .utf8), s.contains("[") { return s }
            return body.removingPercentEncoding
        }
        return t.contains("[Interface]") ? t : nil
    }

    // swiftlint:disable:next large_tuple
    static func parseConf(_ conf: String) -> (String, Int, String, [String: Any])? {
        guard conf.contains("[Interface]") || conf.contains("[Peer]") else { return nil }
        var iface: [String: String] = [:], peer: [String: String] = [:], section = ""
        for raw in conf.split(whereSeparator: \.isNewline) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.caseInsensitiveCompare("[Interface]") == .orderedSame { section = "iface"; continue }
            if line.caseInsensitiveCompare("[Peer]") == .orderedSame { section = "peer"; continue }
            if line.isEmpty || line.hasPrefix("#") || section.isEmpty { continue }
            guard let eq = line.firstIndex(of: "=") else { continue }
            let k = line[..<eq].trimmingCharacters(in: .whitespaces).lowercased()
            let v = line[line.index(after: eq)...].trimmingCharacters(in: .whitespaces)
            if section == "iface" { iface[k] = v } else if peer[k] == nil { peer[k] = v }
        }
        guard let pub = peer["publickey"] else { return nil }
        let endpoint = peer["endpoint"] ?? ""
        var host = endpoint, port = 51820
        if let colon = endpoint.lastIndex(of: ":") {
            host = String(endpoint[..<colon])
            port = Int(endpoint[endpoint.index(after: colon)...]) ?? 51820
        }
        host = host.trimmingCharacters(in: CharacterSet(charactersIn: "[]"))

        var extra: [String: Any] = [:]
        if let v = iface["privatekey"] { extra["private_key"] = v }
        if let v = iface["address"] { extra["local_address"] = v }
        if let v = iface["dns"] { extra["dns"] = v }
        if let v = iface["mtu"] { extra["mtu"] = v }
        if let v = peer["presharedkey"] { extra["preshared_key"] = v }
        if let v = peer["allowedips"] { extra["allowed_ips"] = v }
        if let v = peer["persistentkeepalive"] { extra["keepalive"] = v }
        var found: [String: String] = [:]
        for k in awgKeys { if let v = iface[k] { found[k] = v; extra[k] = v } }
        extra["awg_version"] = version(found)
        return (host, port, pub, extra)
    }

    static func version(_ values: [String: String]) -> String {
        if values.isEmpty { return "plain" }
        if values.keys.contains(where: v3Keys.contains) { return "3.x" }
        if values.keys.contains(where: { ["s3", "s4"].contains($0) }) ||
            values.values.contains(where: { $0.contains("-") }) { return "2.0" }
        if values.keys.contains(where: { $0.hasPrefix("i") }) { return "1.5" }
        return "1.0"
    }
}
